src/application/common/logger.ts

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright (C) 2017 Ericsson and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { inject, injectable, optional } from 'inversify';
+import { LoggerWatcher } from './logger-watcher';
+import { ILoggerServer } from './logger-protocol';
+
+export namespace LogLevel {
+    export const FATAL = 60;
+    export const ERROR = 50;
+    export const WARN = 40;
+    export const INFO = 30;
+    export const DEBUG = 20;
+    export const TRACE = 10;
+}
+
+export const LoggerFactory = Symbol('LoggerFactory')
+export type LoggerFactory = (options?: object) => ILogger;
+
+export const LoggerOptions = Symbol('LoggerOptions')
+
+export const ILogger = Symbol('ILogger');
+
+export interface ILogger {
+    child(obj: Object): ILogger;
+    setLogLevel(logLevel: number): Promise<void>
+    getLogLevel(): Promise<number>;
+    trace(message: string, ...params: any[]): void;
+    debug(message: string, ...params: any[]): void;
+    info(message: string, ...params: any[]): void;
+    warn(message: string, ...params: any[]): void;
+    error(message: string, ...params: any[]): void;
+    fatal(message: string, ...params: any[]): void;
+}
+
+@injectable()
+export class Logger implements ILogger {
+
+    /* Log level for the logger.  */
+    protected _logLevel: Promise<number>;
+
+    /* Root logger has id 1.  */
+    protected readonly rootLoggerId = 1;
+
+    /* Default id is the root logger id.  */
+    protected id: Promise<number> = Promise.resolve(this.rootLoggerId);
+
+    /**
+     * Build a new Logger.
+     *
+     * @param options - The options to build the logger with, see the
+     * bunyan child method documentation for more information.
+     */
+    constructor(
+        @inject(ILoggerServer) protected readonly server: ILoggerServer,
+        @inject(LoggerWatcher) protected readonly loggerWatcher: LoggerWatcher,
+        @inject(LoggerFactory) protected readonly factory: LoggerFactory,
+        @inject(LoggerOptions) @optional() options: object) {
+
+        /* Creating a child logger.  */
+        if (options !== undefined) {
+            this.id = server.child(options);
+        }
+
+        /* Fetch the log level so it's cached in the frontend.  */
+        this.id.then(id => this._logLevel = server.getLogLevel(id));
```

Comment:
should be `this._logLevel = this.id.then(id => this.server.getLogLevel(id))`, otherwise `this._logLevel` can be undefined